xds/src/main/java/io/grpc/xds/ClientLoadCounter.java

```diff
@@ -303,4 +305,33 @@ public void onLoadReport(OrcaLoadReport report) {
       }
     }
   }
+
+  /**
+   * Factory class for creating {@link OrcaPerRequestReportListener} and
+   * {@link OrcaOobReportListener} that record received metric values in the provided
+   * {@link StatsCounter}.
+   */
+  abstract static class XdsMetricsRecordingListenerFactory {
```

Comment:
Have you considered making `StatsCounter` and `XdsMetricsRecordingListenerFactory` final classes, and verifying the side effects in the test? Their implementation/semantics are trivial enough and don't need to be mocked.

Reply:
`XdsMetricsRecordingListenerFactory` is introduced to test the interaction between `LocalityStore` and `OrcaPreRequestUtil`, i.e., at some point of time `LocalityStore` creates a locality-aggregating listener and uses that to create an `OrcaClientStreamTracerFactory` that applies to picker's `PickResult`. How can we test this interaction if `XdsMetricsRecordingListenerFactory` is not mockable?


`StatsCounter` can be made final as there is no external interaction and all its behavior is under [`XdsClientLoadRecorder`](https://github.com/grpc/grpc-java/blob/e795f14bedc56af2f2ba60eef1b23add6106ab99/xds/src/main/java/io/grpc/xds/ClientLoadCounter.java#L255), which is tested. (TODO) Actually we are missing the test for interaction between `XdsClientLoadRecorder` and `XdsLoadStatsStore`, i.e., verify that `XdsLoadStatsStore#interceptPickResult()` applies an `XdsClientLoadRecorder` to the `PickResult`. Again, this requires to introduce another factory (which needs to be mockable) for creating `XdsClientLoadRecorder` instances. Are we introducing too many factories in order to isolate each piece while verifying their interactions? I need some suggestions. Thanks.

Reply:
Update: Actually `ClientLoadCounter`'s implementation, which extends `StatsCounter`, isn't that trivial, especially with the handling for recording backend metrics. There will be too many settings for its internal state if we want to manipulate it in tests. So I prefer keeping `StatsCounter` mockable, which can be easily stubbed for interaction with `StatsStore`.

Reply:
Unless I am missing something, to me `ClientLoadCounter` is as simple as you put some values in it and later get some out as snapshots. The complexity in backend metrics are just for performance, while the semantics are quite simple. Maybe you can point me to an actual test to show me how a final `StatsCounter` is a problem.

Reply:
Let's see [here](https://github.com/grpc/grpc-java/blob/ff33ecd339eacdf3937313015bdddfa4f95dd5c6/xds/src/test/java/io/grpc/xds/XdsLoadStatsStoreTest.java#L234), we want to test 

> Calling method `XdsLoadStatsStore#generateLoadReport()` gives you the current snapshots of all existing locality counters.

If we do not have `StatsCounter` mockable, we need to make real `ClientLoadCounter`s with metric values. In `ClientLoadCounter`'s implementation, we do not have a read-to-use state representation for metric values, so it's not easy to construct them directly. The easiest way I can imagine is to add in metric data by calling `ClientLoadCounter#recordMetric()` individually. So we aren't isolating `XdsLoadStatsStore` from how its internal counters record data.

(Update: seems it's not such a big deal, we can make it final and make test metric values by calling `ClientLoadCounter#recordMetric()`.)

Reply:
I understand. It's a trade-off between how involved the test is and how complex the implementation class hierarchy is. Besides the complexity brought by additional interfaces, we actually try to avoid mocks whenever possible, because mocks can be easily done wrong and can make the test behave differently from real code. That's why we don't allow mockable stubs, and SynchronizationContext is not mockable either. It has to cause us a significant hassle in tests before we start to introduce mockable components. Constructing fake `ClientLoadCounter`s in here doesn't sound a big enough hassle.

I agree this would have some overlap with `ClientLoadCounterTest#normalRecordingOperations()`, and that's fine. `ClientLoadCounterTest#normalRecordingOperations()` is a complete unit test for `ClientLoadCounter` and it tries many different styles of input data, while `XdsLoadStatsStoreTest#loadReportMatchesSnapshots()` can be tested with a `ClientLoadCounter` constructed in the simplest way (e.g., you don't need to add multiple values for a single metric) to get the snapshot value we desire.

Reply:
Alright, #5890 removes `StatsCounter`.